### PR TITLE
8277426: Optimize mask reduction operations on x86

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch improves the performance of mask reduction operations on AVX by matching the pattern `VectorMaskReduction (VectorStoreMask mask)` to eliminate the extra `VectorStoreMaskNode`. I have also done some refactoring to unify the logic of `toLong` with the other reduction operations.

The patch has been discussed partially in [panama-vector repository](https://github.com/openjdk/panama-vector/pull/158).

Thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277426](https://bugs.openjdk.java.net/browse/JDK-8277426): Optimize mask reduction operations on x86

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)
 * [Jie Fu](https://openjdk.java.net/census#jiefu) (@DamonFool - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6447/head:pull/6447` \
`$ git checkout pull/6447`

Update a local copy of the PR: \
`$ git checkout pull/6447` \
`$ git pull https://git.openjdk.java.net/jdk pull/6447/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6447`

View PR using the GUI difftool: \
`$ git pr show -t 6447`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6447.diff">https://git.openjdk.java.net/jdk/pull/6447.diff</a>

</details>

### CODE REVIEW

**Comment by sviswa7** [comment]
@merykitty Thanks for contributing this optimization. The patch looks good to me.

**Comment by sviswa7** [comment]
@PaulSandoz Could you please run this through Oracle testing.

**Comment by PaulSandoz** [comment]
@merykitty does this PR still disable the operations on Neon?

**Comment by sviswa7** [comment]
@PaulSandoz The changes are specific to x86 code gen. Only x86 backend files are changed.

**Comment by PaulSandoz** [comment]
> @PaulSandoz Could you please run this through Oracle testing.

Tests passed.

**Comment by sviswa7** [comment]
Thanks a lot @PaulSandoz for the testing.

**Review by sviswa7 — APPROVED** [review]

**Comment by merykitty** [comment]
@PaulSandoz @sviswa7  Thank you very much for the reviews and testing. Could I get this PR sponsored, pleased?

**Comment by PaulSandoz** [comment]
This needs another hotspot reviewer to review before integration.

**Comment by sviswa7** [comment]
@DamonFool Could you please review this patch?

**Comment by DamonFool** [comment]
> @DamonFool Could you please review this patch?

I will do some testing and feedback.
Thanks.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 1]
```diff
@@ -4061,58 +4061,131 @@ void C2_MacroAssembler::masked_op(int ideal_opc, int mask_len, KRegister dst,
 
 #ifdef _LP64
 void C2_MacroAssembler::vector_mask_operation(int opc, Register dst, KRegister mask,
-                                              Register tmp, int masklen, int masksize,
-                                              int vec_enc) {
+                                              int masklen, int masksize, int vec_enc) {
+  assert(VM_Version::supports_popcnt() &&
```
Comment (DamonFool) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:4065]: New instructions like `lzcntq` and `tzcntq` are used for the optimized code gen without detecting the availability.
I'm a bit worried about that.

So do all AVX512 platforms support them?
Thanks.

**Comment by DamonFool** [comment]
> > @DamonFool Could you please review this patch?
> 
> I will do some testing and feedback. Thanks.

All the tests passed.
But I'm not sure whether all avx512 machines support bmi1 and lzcnt features.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp (hunk 1 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp]: Yes, you are right. I can't find concrete evidence that AVX512 implies BMI1. In addition, `VectorMaskGen` does a check for AVX3 simultaneously with a check for BMI2. So it seems safer to do the same as we did with AVX1 - 2. As a result, I refactored the reduction step into a separate function.
Thank you very much.

**Review by DamonFool — APPROVED** [review]
Thanks for your update.

All my tests passed with the latest version.
So LGTM.

**Comment by merykitty** [comment]
@DamonFool Thanks a lot for your review. Do I need a re-review from @sviswa7  now?

**Comment by DamonFool** [comment]
> @DamonFool Thanks a lot for your review. Do I need a re-review from @sviswa7 now?

Yes, I think so.
Thanks.

**Comment by sviswa7** [comment]
@PaulSandoz Do we need another test run after the changes? If so, could you please help in running through your testing.

**Comment by PaulSandoz** [comment]
@sviswa7 re-running tests since there were some subtle changes in the code

**Comment by PaulSandoz** [comment]
Latest tests are good.

**Comment by DamonFool** [comment]
> Latest tests are good.

So let's integrate it.
Thanks.

/sponsor

**Pull request closed without merging**